velox/functions/lib/LikeFunctionUtils.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/lib/LikeFunctionUtils.h"
+#include "velox/vector/TypeAliases.h"
+
+namespace facebook::velox::functions {
+namespace likeutils {
+
+bool isWildcardInPattern(const StringView& pattern) {
+  vector_size_t patternLength = pattern.size();
+  auto patternStr = pattern.data();
+  for (vector_size_t i = 0; i < patternLength; i++) {
+    if ((patternStr[i] == '%') || (patternStr[i] == '_')) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool matchExactPattern(const StringView& input, const StringView& pattern) {
+  vector_size_t inputLength = input.size();
+  vector_size_t patternLength = pattern.size();
+
+  if (inputLength != patternLength) {
+    return false;
+  }
+  return (std::memcmp(input.data(), pattern.data(), inputLength) == 0);
+}
+
+vector_size_t isPrefixPattern(const StringView& pattern) {
+  vector_size_t patternLength = pattern.size();
+  vector_size_t i = patternLength - 1;
+  auto patternStr = pattern.data();
+
+  for (; i >= 0; i--) {
+    if (patternStr[i] != '%') {
+      break;
+    }
+  }
+  if (i == patternLength - 1) {
+    return 0;
+  }
+  auto prefixLength = i + 1;
+
+  for (; i >= 0; i--) {
+    if ((patternStr[i] == '%') || (patternStr[i] == '_')) {
+      return 0;
+    }
+  }
+  return prefixLength;
+}
+
+bool matchPrefixPattern(
+    const StringView& input,
+    const StringView& pattern,
+    vector_size_t length) {
+  return (
+      std::memcmp(
+          input.getString().substr(0, length).c_str(),
+          pattern.getString().substr(0, length).c_str(),
+          length) == 0);
+}
+
+vector_size_t isSuffixPattern(const StringView& pattern) {
+  vector_size_t i = 0;
+  vector_size_t patternLength = pattern.size();
+  auto patternStr = pattern.data();
+
+  for (; i < patternLength; i++) {
+    if (patternStr[i] != '%') {
+      break;
+    }
+  }
+  if (i == 0) {
+    return 0;
+  }
+  auto suffixLength = patternLength - i;
+
+  for (; i < patternLength; i++) {
+    if ((patternStr[i] == '%') || (patternStr[i] == '_')) {
+      return 0;
+    }
+  }
+  return suffixLength;
+}
+
+bool matchSuffixPattern(
+    const StringView& input,
+    const StringView& pattern,
+    vector_size_t length) {
+  auto inputStr = input.getString();
+  auto patternStr = pattern.getString();
+  std::reverse(inputStr.begin(), inputStr.end());
```

Comment:
This is not necessary. We just need to check that input.size() >= length and then

`std::memcmp(input.data() + input.size() - length, pattern.data() + pattern.size() - length, length);`

Reply:
Thanks for sharing the simplified code!